Checkout: Update `thank-you` page route

drewblaisdell opened the pull request:
Before #3270 is merged, the API needs to be updated to redirect users to the new thank you page route at `/checkout/:site/:receipt_id/thank-you`. Before the API can be updated to redirect users to this new route, this route needs to work in Calypso.

This PR updates `Checkout` and the upgrades index to render the thank you page on `/checkout/:site/:receipt_id/thank-you`. This route works for any value of `:site` or `:receipt_id`, as the checkout process is still reliant on data in local storage. This is necessary to prevent errors for users before or after the patch that this PR depends on is deployed.

That patch is here: D1083-code
**Note:** You must point `foobar.wordpress.com` at your sandbox's IP in your hosts file, where *foobar* is the site slug of the site you are purchasing a plan (or other item) for when testing.

**Testing**
We need to assert that it is possible to reach the thank you page with and without the patch applied, with a credit card or with PayPal.

*Without the patch applied*
- Purchase an item (e.g. a plan) with a credit card.
- Assert that you land on the thank you page at `/checkout/:site/thank-you`, where `:site` is the slug of the site you checked out for.
- Purchase an item (e.g. a plan) with PayPal.
- Assert that you land on the thank you page at `/checkout/:site/thank-you`, where `:site` is the slug of the site you checked out for.

*With the patch applied*
- Purchase an item (e.g. a plan) with a credit card.
- Assert that you land on the thank you page at `/checkout/:site/thank-you`, where `:site` is the slug of the site you checked out for.
- Purchase an item (e.g. a plan) with PayPal.
- Assert that you land on the thank you page at `/checkout/:site/:receipt_id/thank-you`, where `:site` is the slug of the site you checked out for and `:receipt_id` is the ID of the receipt that was emailed to you.

- [x] Code review
- [x] Product review

stephanethomas commented:
> Before the API can be updated to redirect users to this new route, this route needs to work in Calypso.

We could also bump the version number of the `/me/transactions` API endpoint to make sure clients use the appropriate version.

stephanethomas commented on client/my-sites/upgrades/checkout/checkout.jsx:
```diff
@@ -133,7 +133,7 @@ const Checkout = React.createClass( {
 		}
 
-		return '/checkout/thank-you';
+		return `/checkout/${ this.props.sites.getSelectedSite().slug }/:receipt_id/thank-you`;
```
This should be `:receiptId` rather than `:receipt_id`.

drewblaisdell replied:
Thanks for catching this.

scruffian commented:
Code looks good to me

stephanethomas commented:
Tested and works fine despite the wrong path fragment. Thanks for the detailed testing instructions!

stephanethomas commented:
This should be good to go.

drewblaisdell commented:
I'm moving this back to 'In Progress' while I add a route for a generic thank you page, which we need in case the API redirects users back without a `:receiptId`.

drewblaisdell commented:
Actually, the generic thank you page will be added in #3270, because the codebase still relies on local storage with this PR. I updated this PR to:

- Make the `/:receiptId` portion of the path optional, for the case where the API doesn't obtain a receipt ID and redirects to `/checkout/:site/thank-you`.
- Redirect users to `/checkout/:site/thank-you` when checking out with a credit card. This ensures that users won't see `:receiptId` in the address bar, even temporarily.

I updated the testing instructions and unchecked both the product and code boxes.

scruffian commented:
LGTM :+1:
